### PR TITLE
✨ Feat/save cfg

### PR DESCRIPTION
**Pull request by loukotal**
## WDTPRD?
- Allows for saving tokens to config
- Config gets created in `.config/mrdog.toml`
  - probably needs checks for `.config`, that is currently missing, but it is what it is
- Allows for having multiple tokens per provider
  - We probably would want to not save the provider with the values, maybe there's some "serde exclude"?

- [x] There's also bunch of warnings, sorry for the noise in the recording! I'll add commit removing them 

## Vid

https://user-images.githubusercontent.com/18210858/219867227-83e90ad8-6348-4468-bfd3-0f4fb729314c.mov

## Obligatory meme
7bmeyv

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -1,6 +1,9 @@
+use std::error::Error;
+
 mod cli;
 mod mrdog;
 
-fn main() {
-    cli::run();
+fn main() -> Result<(), Box<dyn Error>> {
+    cli::run()?;
+    Ok(())
```
Comment (smolijar) [inline src/main.rs:8]: NIT: Typically, the main fn does not have an error result (nobody process it anyway), so it returns void and panics. You can achieve this by `run().unwrap()` I think.

File: src/cli.rs [hunk 2]
```diff
@@ -21,13 +24,25 @@ pub struct Cli {
...
-        },
+        Some(Commands::Set {
+            value,
+            name,
+            provider,
+        }) => {
+            let config_value = ConfigValue {
+                value,
+                name,
+                provider,
```
Comment (smolijar) [inline src/cli.rs:38]: NIT: Now we have a "global" option _provider_, which will look funny once we use the same "config" to set e.g. defualt view, etc.

File: src/cli.rs [hunk 3]
```diff
@@ -1,17 +1,20 @@
...
-
-#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, ValueEnum)]
-pub enum ConfigValue {
-    Gitlab,
-    Github,
+    Set {
+        // TODO: allow shorthand values like gh / gl
+        #[arg(long = "provider", default_value_t = Provider::Github, value_enum)]
+        provider: Provider,
+        #[arg(short = 'n', default_value_t = String::from("default"))]
```
Comment (smolijar) [inline src/cli.rs:14]: I love that you added names for the tokens! :heart:

File: src/mrdog.rs [hunk 4]
```diff
@@ -1,19 +1,33 @@
...
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct ConfigValue {
+    pub name: String,
+    pub value: String,
+    pub provider: Provider,
 }
 
-pub fn set_config(item: ConfigItem) {
-    config_storage::set_config(item)
+pub fn set_config(item: ConfigValue) -> Result<(), Box<dyn Error>> {
```
Comment (smolijar) [inline src/mrdog.rs:24]: NIT: Ideally, instead of dyn/sad we'd have specific error enum: FileNotFound, InvalidFormat, etc.

File: src/mrdog.rs [hunk 5]
```diff
@@ -1,19 +1,33 @@
...
+    let c = config_storage::set_config(item).map_err(|_| "sad")?;
+    println!("{c:?}");
+    Ok(())
 }
 
-pub fn list() {
+pub fn list() -> Result<(), Box<dyn Error>> {
```
Comment (smolijar) [inline src/mrdog.rs:30]: NIT: This is not related to the config, right?

File: src/mrdog/config_storage.rs [hunk 6]
```diff
@@ -1,5 +1,93 @@
...
+pub enum ConfigError {
+    CreateFailed(io::Error),
+    CouldNotConvertToToml(toml::ser::Error),
+    WriteToConfigFailed(io::Error),
+}
+
+#[derive(Deserialize, Serialize, Debug)]
+struct Config {
+    github: Option<Vec<ConfigValue>>,
+    gitlab: Option<Vec<ConfigValue>>,
```
Comment (smolijar) [inline src/mrdog/config_storage.rs:20]: NIT: Now this is when the structure I drafted became retarded, I would expect here not generic `ConfigValue` but the GL token directly (or wrappign structure). It is retardness we know about and will change ruthlessly in the future.

File: src/mrdog/config_storage.rs [hunk 7]
```diff
@@ -1,5 +1,93 @@
...
+
+#[derive(Debug)]
+pub enum ConfigError {
+    CreateFailed(io::Error),
+    CouldNotConvertToToml(toml::ser::Error),
+    WriteToConfigFailed(io::Error),
+}
+
+#[derive(Deserialize, Serialize, Debug)]
+struct Config {
```
Comment (smolijar) [inline src/mrdog/config_storage.rs:18]: Now that I see it, adding "version" might be a good idea when someone starts using except myself and my mom.

File: src/mrdog/config_storage.rs [hunk 8]
```diff
@@ -1,5 +1,93 @@
...
+    let written = fs::File::create(get_config_path());
+    return written
+        .map(|_| true)
+        .map_err(|e| ConfigError::CreateFailed(e));
+}
+
+fn get_config_path() -> PathBuf {
+    let home = dirs::home_dir();
+    if let Some(p) = home {
+        p.join(".config/mrdog.toml")
```
Comment (smolijar) [inline src/mrdog/config_storage.rs:89]: NIT: In my expirience, on Linux peeps tend to create a subfolder, e.g. `.config/mrdog/config.toml` to put all future trash in subfolder.

**Review by smolijar — APPROVED** [review]
Few comments, consider all NITs. We shan't block no change request by no lame reviews. Mrdog na to & lets roll!

Thanks for the PR!

File: src/mrdog.rs (hunk 5 above)
Comment (loukotal) [reply src/mrdog.rs:30]: Nope, this was to remove compiler warnings

File: src/main.rs (hunk 1 above)
Comment (loukotal) [reply src/main.rs:8]: I see, I thought I saw somewhere the `-> Result<>`. Here it seems like an option https://doc.rust-lang.org/rust-by-example/error/result.html?highlight=main#using-result-in-main. Seems like it should print out the error using the `Debug` trait (though no clue what happens for this case I made here 😅)

File: src/mrdog/config_storage.rs (hunk 8 above)
Comment (loukotal) [reply src/mrdog/config_storage.rs:89]: I've checked mine, and there were both files and subfolders :D Will add a subfolder 🐱

Comment (loukotal) [reply src/mrdog/config_storage.rs:89]: > I've checked mine

reminded me of this gem :D https://www.youtube.com/watch?v=c8AP4qDBfqo

**Comment by loukotal** [comment]
LFG, i'm merging. Thanks for the review! I will go through the NITs in next PR(s)
